test/test_processes.py

```diff
@@ -626,7 +629,30 @@ class ParentProcess(plumpy.Process):
             def run(self):
                 StackTest().execute()
 
+        reentrant_on()
```

Comment:
Shouldn't we just make this the default in the `setUp` of the test case?

Reply:
Well, if I understand it correctly, we only turn on the reentrant setup when we actually need it, just like this specific test case. But for other cases, we keep using the regular event loop.

Reply:
I see what you're saying, but my concern is that for making `plumpy` easy to use, I feel like the instructions we should give to people is 'If you use any part of plumpy, always turn reentrant on before you do anything`.  This way, people don't have to think about it (or know the details of why we have this).  They just turn in on just as their code is about to start calling plumpy methods/classes and forget about it.  
What do yout think?

Reply:
I agree this should be hidden from users. And I guess the proper place to put this is right before `proc.execute()`. 

But I find I make something not correct in this PR. The loop instance is not consistent before and after 'reentrant'.   You can check the following snippet:class MyEventLoopPolicy(asyncio.DefaultEventLoopPolicy):

    print(id(asyncio.get_event_loop()))


    def get_event_loop(self):
    """Get the event loop.

          This may be None or an instance of EventLoop.
    """

    loop = super().get_event_loop()
    print(id(loop))
     # Do something with loop ...

    return loop

asyncio.set_event_loop_policy(MyEventLoopPolicy())
```
class MyEventLoopPolicy(asyncio.DefaultEventLoopPolicy):

    print(id(asyncio.get_event_loop()))


    def get_event_loop(self):
    """Get the event loop.

          This may be None or an instance of EventLoop.
    """

    loop = super().get_event_loop()
    print(id(loop))
     # Do something with loop ...

    return loop

asyncio.set_event_loop_policy(MyEventLoopPolicy())
```

I assume the loop should be the same one, but it is not.

Reply:
Hi @muhrin , could you have a look at this. In my opinion, this is an exceptional behaviour that prevents us from keep on using one loop before and after reentrant.

Reply:
Yes, you're right!  I'm sorry I didn't think about this when I looked at it.

What I think should be done is that the event loop should be lazily created, stored, and then the same instance be given out each time, e.g.:


```
class MyEventLoopPolicy(asyncio.DefaultEventLoopPolicy):

    def __init__(self):
        super().__init__()
        self._loop = None

    def get_event_loop(self):
        if self._loop is None:
            self._loop = super().get_event_loop()

        return self._loop
```
(untedted)

Maybe you see what I'm getting at from this example.  Basically, that it's the job of our policy to make sure that only one loop is ever created and that the same one is always returned.

The more I think about it, the more I feel that `reentrant_on/off` should just set or remove the policy, and it's the policy (in `new_event_loop()`) that should apply the reentrancy to the loop the first time we create it.

Reply:
> The more I think about it, the more I feel that reentrant_on/off should just set or remove the policy, and it's the policy (in new_event_loop()) that should apply the reentrancy to the loop the first time we create it.

I think this makes a lot of sense indeed. How should the custom policy be loaded automatically. Should it just be put in `plumpy/__init__.py`? Is there a performance hit for executing the `asyncio.set_event_loop_policy(PlumpyEventLoopPolicy())` multiple times? Also, where should the "reset" code go? Should that just remain a function in `plumpy`? Maybe just `reset_policy` instead of `reentrant_off`? That decouples the name a bit from what actually happens and is more future proof I think

Reply:
Thanks Seb.  We discussed some of this with @unkcpz and I'm of the opinion that making global changes (particularly as consequential as setting `asyncio`'s loop policy) on import is bad etiquette and leads to the problem of not being able to import `plumpy` and possibly `aiida` without nuking your current policy (even if you don't intend to run anything with AiiDA).  Unfortunately I can only see one alternative, which is to make it an explicit call, either on runner creation or runner execute (perhaps the latter is better).  In this case, the 'reset' could happen after the runner finishes executing.

As for the names, yes, I like the suggestion.  There's no reason a casual user needs to know we're doing this for reentrancy, but they might want to know that we're overwriting the global policy so `set/reset_policy` or similar might work well.

Reply:
Well, things going like this: every new event policy instant will have different results of `get_event_loop()`. 
You see:
```
print('t11', id(asyncio.DefaultEventLoopPolicy().get_event_loop()))
print('t11', id(asyncio.DefaultEventLoopPolicy().get_event_loop()))
```
not the same loop. Even though the current loop is set to the thread. This lead to that `super(),get_event_loop()` return a new event loop.

Reply:
Of course, that makes sense.  But what I'm proposing is that a user that we cache the loop in the policy object and the users only sets the policy once, so they _should_ get the same loop back each time by virtue of this logic:

```
        if self._loop is None:
            self._loop = super().get_event_loop()

        return self._loop
```

Reply:
But once we set the new policy the `asyncio.get_event_loop()` return another loop? What I understand is that no matter when user calls `asyncio.get_event_loop()`, same event loop returned.

Reply:
I'm not sure I quite understand.  

What I think we want is a policy that the first time that it is used after being set creates one new event loop and makes it reentrant.  After that point, `get_event_loop()` and `new_event_loop()` (btw, this is something we need to add to the policy) only ever return that same (reentrant) loop instance.  Any loops created or used, manually or automatically (through, e.g., `asyncio.get_event_loop()`) _before_ the policy is registered are not changed and are not used by AiiDA/plumpy.

Reply:
I see. I just too obsessed with the same loop before and after the custom loop policy.
```
[loop{A1}] ----> [loop{B1}] -----> [{loop{A2}}]       (current)       

[loop{A1}] ----> [loop{B1}] -----> [{loop{A1}}]       (I expected) 
```
But it seems it is not necessary(and useless) to have the loop exactly the one before the custom policy is setting.